### PR TITLE
mon,osd: add no{out,down,in,out} flags on CRUSH nodes

### PR DESCRIPTION
**Pull request by liewegas**
- put these in a new OSDMap crush_node_flags map that can take CEPH_OSD_*
- in practice, only the no* flags are used here
- re-use the existing 'ceph osd add-no*' and 'ceph osd rm-no*' commands
- re-use the existing OSD_FLAGS health warning, but update the docs to make it generalize to crush nodes (not just OSDs)

### CODE REVIEW

File: src/osd/OSDMap.cc [hunk 1]
```diff
@@ -5784,3 +5843,20 @@ float OSDMap::pool_raw_used_rate(int64_t poolid) const
     ceph_abort_msg("unrecognized pool type");
   }
 }
+
+unsigned OSDMap::get_crush_node_flags(int osd) const
+{
+  unsigned flags = 0;
+  if (!crush_node_flags.empty()) {
+    // the map will contain type -> name
+    std::map<std::string,std::string> ploc = crush->get_full_location(osd);
```
Comment (xiexingguo) [inline src/osd/OSDMap.cc]: What if we have different flags for different parents? E.g.,
```
rack-1              {no-out, no-down}
host-1              {no-out, no-in}
```
or we should not allow that to happen at all..

**Comment by xiexingguo** [comment]
@liewegas 
Since you are touching these codes, I wonder if we can add an `admin down` command too:
- `admin down` can be used to force stopping a specified osd(s)
- `admin down` osd(s) won't be automatically marked out by Ceph

File: src/osd/OSDMap.cc (hunk 1 above)
Comment (liewegas) [reply src/osd/OSDMap.cc]: in that case, if host-1 is in rack-1, an osd on host-1 would have all 3 flags set...

**Comment by liewegas** [comment]
ceph osd admin-down <osd.nnn> ?

**Comment by xiexingguo** [comment]
> ceph osd admin-down <osd.nnn> ?
YES.

> I like that, but we need to pick a way for the osdmap to indicate that the osd should exit and stop on its own

The easiest way would be introduce a new per osd CEPH_OSD_ADMIN[FORCE]_OUT flag. Then `_committed_osd_maps`  should be able to take good care of it.


Would you like me to give it a try? Or you intend to do it yourself here :-)

**Comment by xiexingguo** [comment]
> ceph osd admin-down <osd.nnn> ?

YES.

> I like that, but we need to pick a way for the osdmap to indicate that the osd should exit and stop on its own

The easiest way would be introducing a new per osd CEPH_OSD_ADMIN[FORCE]_DOWN flag. Then `_committed_osd_maps`  should be able to take good care of it.


Would you like me to give it a try? Or you intend to do it yourself here :-)

**Comment by xiexingguo** [comment]
@liewegas See https://github.com/ceph/ceph/pull/27595

**Pull request merged**